File: android/GeminiBluetoothBridge/app/src/main/java/com/example/geminibridge/BleConstants.kt
package com.example.geminibridge

import java.util.UUID

object BleConstants {
    val serviceUuid: UUID = UUID.fromString("8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c01")
    val writeCharUuid: UUID = UUID.fromString("8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c02")
    val notifyCharUuid: UUID = UUID.fromString("8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c03")
    val cccdUuid: UUID = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")

    const val protocolVersion: Int = 1
    const val frameHeaderBytes: Int = 7
    const val defaultAttMtu: Int = 23
    const val defaultMaxPacketSize: Int = 20
    const val maxGattAttributeValueBytes: Int = 512
    const val assemblyTimeoutMs: Long = 300_000L

    // Protocol v2 binary ping/pong (8 bytes vs ~55 bytes JSON — 85% smaller)
    // Frame: [0xFE][0xFD][type:1][ts_ms:4 big-endian][status:1]
    val binaryFrameMagic: ByteArray = byteArrayOf(0xFE.toByte(), 0xFD.toByte())
    const val binaryPingType: Byte = 0x01
    const val binaryPongType: Byte = 0x02
    const val binaryFrameSize: Int = 8 // magic(2) + type(1) + ts_ms(4) + status(1)

    fun isBinaryFrame(data: ByteArray): Boolean =
        data.size >= binaryFrameSize && data[0] == binaryFrameMagic[0] && data[1] == binaryFrameMagic[1]

    fun buildBinaryPong(tsMs: Long): ByteArray {
        val buf = java.nio.ByteBuffer.allocate(binaryFrameSize).order(java.nio.ByteOrder.BIG_ENDIAN)
        buf.put(binaryFrameMagic[0])
        buf.put(binaryFrameMagic[1])
        buf.put(binaryPongType)
        buf.putInt((tsMs and 0xFFFFFFFFL).toInt())
        buf.put(0)
        return buf.array()
    }

    fun parseBinaryFrameType(data: ByteArray): Byte = data[2]
}

File: GeminiAirBridge-iOS/Sources/GeminiAirBridge/BLE/BLEConstants.swift
import Foundation
import CoreBluetooth

// MARK: - GATT Service UUIDs
/// Matches Android BleConstants exactly: 8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c0{1,2,3}
public enum BLEConstants {
    public static let serviceUUID      = CBUUID(string: "8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c01")
    public static let writeCharUUID    = CBUUID(string: "8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c02")
    public static let notifyCharUUID   = CBUUID(string: "8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c03")
    public static let cccdUUID         = CBUUID(string: "00002902-0000-1000-8000-00805f9b34fb")

    // MARK: - Frame protocol constants
    public static let protocolVersion: UInt8      = 1
    public static let frameHeaderBytes: Int        = 7
    public static let defaultAttMtu: Int           = 23
    public static let defaultMaxPacketSize: Int    = 20
    public static let maxGattAttributeValueBytes   = 512
    public static let assemblyTimeoutSeconds: TimeInterval = 300.0

    // MARK: - Binary ping/pong v2 (8 bytes, 85% smaller than JSON)
    /// Frame: [0xFE][0xFD][type:1][ts_ms:4 big-endian][status:1]
    public static let binaryFrameMagic: Data = Data([0xFE, 0xFD])
    public static let binaryPingType: UInt8  = 0x01
    public static let binaryPongType: UInt8  = 0x02
    public static let binaryFrameSize: Int   = 8

    // MARK: - Prompt bundle magic
    public static let promptBundleMagic: Data = Data("bgp2".utf8)
    public static let promptBundleHeaderBytes = 13
    public static let promptBundleFlagGzipMetadata: UInt8 = 0x01

    // MARK: - Gzip compression
    public static let gzipMagic: Data = Data("gz".utf8) + Data([0x01])
    public static let jsonGzipThresholdBytes = 900

    // MARK: - Heartbeat
    public static let pingIntervalSeconds: TimeInterval = 30.0
    public static let pingTimeoutSeconds: TimeInterval  = 90.0

    // MARK: - Transport protocol
    public static let maxTransportId: UInt16 = 0xFFFF
    public static let maxTotalChunks: UInt16 = 0xFFFF

    // MARK: - Request limits
    public static let maxPromptBytes           = 220 * 1024
    public static let maxImageBytes            = 1_200_000
    public static let maxContextBlocks          = 12
    public static let maxMemoryTurns            = 24
    public static let maxConcurrentPrompts      = 2
    public static let maxPendingPrompts         = 10

    // MARK: - Send throttle
    public static let throttleMultiClientEvery = 6
    public static let throttleSingleClientEvery = 4
    public static let throttleDelayMs: UInt64   = 1

    // MARK: - Convenience
    public static let packetServiceUUIDString = "8e7f1f10-6c7a-4a89-b2e8-4e20f4f31c01"
}

// MARK: - Binary frame helpers
public extension BLEConstants {
    static func isBinaryFrame(_ data: Data) -> Bool {
        data.count >= binaryFrameSize
            && data[0] == binaryFrameMagic[0]
            && data[1] == binaryFrameMagic[1]
    }

    static func parseBinaryFrameType(_ data: Data) -> UInt8 {
        data[2]
    }

    static func buildBinaryPong(tsMs: Int64) -> Data {
        var buf = Data(capacity: binaryFrameSize)
        buf.append(binaryFrameMagic)
        buf.append(binaryPongType)
        withUnsafeBytes(of: UInt32(truncatingIfNeeded: tsMs).bigEndian) { buf.append(contentsOf: $0) }
        buf.append(0)
        return buf
    }

    static func parseBinaryTimestamp(_ data: Data) -> Int64 {
        let raw = UInt32(bigEndian: data.withUnsafeBytes { $0.load(fromByteOffset: 3, as: UInt32.self) })
        return Int64(raw)
    }
}